### ISSUE TITLE
 multi scale training

### ISSUE THREAD

**Issue by scucmpunk** [Junjue-Wang/LoveDA#46]
Hi, I want to know how to make multi scale training?

**Comment by Junjue-Wang** [comment]
You can add muti-scale augmentations into the data transforms.

**Comment by scucmpunk** [comment]
Thank you very much, but I want to know the process further. Does multi-scale training use the Class Scale in the tta.py? Change as follows:
```
OneOf([
            Scale(True)
                ], p=0.75),
```
Is this correct？
 On the other hand, after the training is completed, the following figure will often appear in the prediction results, and the accuracy rate is only about 0.03%. Is this normal?
![ZTB (V8(R4B9EI838S9Q2 G](https://user-images.githubusercontent.com/113426059/190145167-bd4979c2-004b-482a-9dff-879e24930cb8.png)

**Comment by Junjue-Wang** [comment]
No, it seems not normal. TTA refers to test time augmentation. It can not be used in training. You can use some transforms repo such as Albumentations.

**Comment by scucmpunk** [comment]
The picture above is the result of hrnet network after using multi_binary_loss, which looks really bad. As for multi-scale training, I understand. Thank you!

**Comment by scucmpunk** [comment]
Excuse me, Are these also pretrained weights? Is "hrnetv2_w32-dc9eeb4f.pth" different from the pretrained weight provided in README?
model_urls = {
    "hrnetv2_w18": "https://download.openmmlab.com/pretrain/third_party/hrnetv2_w18-00eb2006.pth",
    "hrnetv2_w32": "https://download.openmmlab.com/pretrain/third_party/hrnetv2_w32-dc9eeb4f.pth",
    "hrnetv2_w40": "https://download.openmmlab.com/pretrain/third_party/hrnetv2_w40-ed0b031c.pth",
    "hrnetv2_w48": "https://download.openmmlab.com/pretrain/third_party/hrnetv2_w48-d2186c55.pth",
}

**Comment by Junjue-Wang** [comment]
Please note that they are ImageNet pretrained weights

**Comment by scucmpunk** [comment]
If I want to train hrnetv2_w40, is the pretrained weight unnecessary?

**Comment by Junjue-Wang** [comment]
I think it is necessary because it is the mainstream trick.

**Comment by scucmpunk** [comment]
Excuse me, The test results still appear above, I do not know where there is a problem. I added the test set data loading in loveda.py file, I do not know if there is a problem.
```
test_test=dict(
        type='LoveDALoader',
        params=dict(
            image_dir=[
                './LoveDA/Test/Urban/images_png/',
                './LoveDA/Test/Rural/images_png/',
            ],

            transforms=Compose([
                Normalize(mean=(123.675, 116.28, 103.53),
                          std=(58.395, 57.12, 57.375),
                          max_pixel_value=1, always_apply=True),
                er.preprocess.albu.ToTensor()

            ]),
            CV=dict(k=10, i=-1),
            training=False,
            batch_size=4,
            num_workers=0,
        ),
    ),
```
And added TTA for the test
```
def predict_test(ckpt_path, config_path='base.hrnetw32', save_dir='', use_TTA=True):
    os.makedirs(save_dir, exist_ok=True)
    cfg = import_config(config_path)
    statedict = torch.load(ckpt_path, map_location=lambda storage, loc: storage)
    model_state_dict = remove_module_prefix(statedict)
    print('Load model!')
    test_dataloader = make_dataloader(cfg['data']['test_test'])
    model = make_model(cfg['model'])
    model.load_state_dict(model_state_dict, strict=False)
    model.cuda()
    model.eval()
    TTA = TestTimeAugmentation(model, tta_config=[
                    Scale(scale_factor=0.5),
                    Scale(scale_factor=0.75),
                    Scale(scale_factor=1.0),
                    Scale(scale_factor=1.25),
                    Scale(scale_factor=1.5),
                    Scale(scale_factor=1.75),
                ])
    with torch.no_grad():
        for img, gt in tqdm(test_dataloader):
            img = img.cuda()
            if use_TTA:
                pred = TTA(img)
            else:
                pred = model(img)
            pred = pred.argmax(dim=1).cpu()
            for clsmap, imname in zip(pred, gt['fname']):
                imsave(os.path.join(save_dir, imname), clsmap.cpu().numpy().astype(np.uint8))
    torch.cuda.empty_cache()
```

**Comment by scucmpunk** [comment]
In predict.py, I also modified
```
model.load_state_dict(model_state_dict, strict=False)
```
I don't know if this is the reason

**Comment by scucmpunk** [comment]
The test results with the given pretrained model are normal. Using my own training model, the result is abnormal.

**Comment by Junjue-Wang** [comment]
> In predict.py, I also modified
> 
> ```
> model.load_state_dict(model_state_dict, strict=False)
> ```
> 
> I don't know if this is the reason

If the pretrained weights and model params do not match. It does happen.

**Comment by scucmpunk** [comment]
Why the size of my final trained model is 226.60MB and the size of the pre-trained model is 113.58MB, where should I load the pre-trained model?

**Comment by scucmpunk** [comment]
In the UDA experiment, if I want to predict_test images, it seems that I have to set mask tags?

**Comment by Junjue-Wang** [comment]
You can set None

> In the UDA experiment, if I want to predict_test images, it seems that I have to set mask tags?

**Comment by scucmpunk** [comment]
Why the size of my final trained model is 226.60MB and the size of the pre-trained model is 113.58MB, where should I load the pre-trained model? Please take a look at this question again, thank you!

**Comment by Junjue-Wang** [comment]
Both are ok. Pytorch save the optim params as well as model params you know.
But we often abandom the optim params to slim the model.

**Comment by scucmpunk** [comment]
But there will be such problems
![U3_{2PYGL}A~6N@RUY8KT1Z](https://user-images.githubusercontent.com/113426059/194034970-d95e7a9c-7035-414f-aca5-aa3d9ee3853a.png)

**Comment by Junjue-Wang** [comment]
You should align thsese keys

**Comment by scucmpunk** [comment]
Thank you for your advice

**Comment by scucmpunk** [comment]
Hello, the following errors occurred in my rural test. Do you know how to solve them？
![2{)~WC%5F5DXF2CT}9~FE$I](https://user-images.githubusercontent.com/113426059/196205494-c6e28b82-47f3-42b0-9526-1696237ad73c.png)

**Comment by Junjue-Wang** [comment]
Please check your submitted ids.

**Comment by scucmpunk** [comment]
But the rural test set is from 4191 to 5166 and I don‘t see the relationship with 2747 in the figure.

**Comment by Junjue-Wang** [comment]
![image](https://user-images.githubusercontent.com/27996790/196437844-9fc5c1ab-e83a-4b1f-baf1-4e16b571fecc.png)
No. Test rural images only have 976 images

**Comment by scucmpunk** [comment]
Can you help me to find out why I can't log on to the competition website now?

**Comment by scucmpunk** [comment]
How do I set up multi-GPU training?

**Comment by sodaaoy** [comment]
@scucmpunk @Junjue-Wang ,oh! I tried to get the  hrnetv2_w40_imagenet_pretrained.pth in many ways, is this file in your comment "hrnetv2_w40": "https://download.openmmlab.com/pretrain/third_party/hrnetv2_w40-ed0b031c.pth" the same to hrnetv2_w40_imagenet_pretrained.pth?
